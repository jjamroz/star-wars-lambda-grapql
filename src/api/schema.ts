import { gql } from 'apollo-server-lambda';

export const typeDefs = gql`
  type Character {
    id: String!
    name: String!
    episodes: [String!]!
    planet: String
  }

  type AllCharacterPaginated {
    items: [Character!]!
    lastEvaluatedKey: String
  }

  input CharacterInput {
    name: String!
    episodes: [String!]!
    planet: String
  }

  input PatchCharacterData {
    name: String
    episodes: [String!]
    planet: String
  }

  input UpdateCharacterInput {
    id: String!
    patch: PatchCharacterData!
  }

  type Query {
    getAllCharacters(limit: Int, lastEvaluatedKey: String): AllCharacterPaginated!
    getCharacter(id: String!): Character
  }
  type Mutation {
    createCharacter(input: CharacterInput!): Character
    updateCharacter(input: UpdateCharacterInput!): Character
    deleteCharacter(id: String!): Character
  }
`;
