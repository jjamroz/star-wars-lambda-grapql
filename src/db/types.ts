export interface Entity {
  id: string;
}

export interface Character extends Entity {
  name: string;
  episodes: string[];
  planet?: string | null;
}

export type CreateDto<T> = Omit<T, 'id'>;
export type GetDto<T extends Entity> = Pick<T, 'id'>;
export type DeleteDto<T extends Entity> = Pick<T, 'id'>;
export type UpdateDto<T extends Entity> = {
  id: T['id'];
  patch: Partial<Omit<T, 'id'>>;
};

export type PaginatedData<T extends Entity> = {
  items: [T];
  lastEvaluatedKey: Entity;
};

export type PaginatedSearch = {
  limit?: number;
  lastEvaluatedKey?: Entity['id'];
};
