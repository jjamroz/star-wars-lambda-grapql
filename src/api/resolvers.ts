import { getCharacter } from './resolvers/get-character.query';
import { getAllCharacters } from './resolvers/get-all-characters.query';
import { createCharacter } from './resolvers/create-character.mutation';
import { updateCharacter } from './resolvers/update-character.mutation';
import { deleteCharacter } from './resolvers/delete-character.mutation';

export const resolvers = {
  Query: {
    getAllCharacters,
    getCharacter,
  },
  Mutation: {
    createCharacter,
    updateCharacter,
    deleteCharacter,
  },
};
