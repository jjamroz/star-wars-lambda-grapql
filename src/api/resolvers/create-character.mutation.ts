import { Character, CreateDto } from '../../db/types';
import { CharacterDynamodb } from '../../db/character-dynamodb';

type createCharacterArgs = {
  input: CreateDto<Character>;
};

export const createCharacter = async (
  _: any,
  { input }: createCharacterArgs,
): Promise<Character | null> => {
  console.log(`[MUTATION] createCharacter called with ${JSON.stringify(input, null, 4)}`);
  return CharacterDynamodb.Instance.createItem(input);
};
