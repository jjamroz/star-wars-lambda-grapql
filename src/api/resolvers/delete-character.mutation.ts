import { CharacterDynamodb } from '../../db/character-dynamodb';
import { Character, DeleteDto } from '../../db/types';

type deleteCharacterArgs = DeleteDto<Character>;

export const deleteCharacter = async (
  _: any,
  { id }: deleteCharacterArgs,
): Promise<Character | null> => {
  console.log(`[MUTATION] deleteCharacter called with id ${id}`);
  return CharacterDynamodb.Instance.deleteItem<Character>({ id });
};
