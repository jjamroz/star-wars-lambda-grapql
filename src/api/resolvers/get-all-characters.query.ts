import { Character, Entity, PaginatedData } from '../../db/types';
import { CharacterDynamodb } from '../../db/character-dynamodb';

type getAllArgs = {
  limit: number;
  lastEvaluatedKey: Entity['id'];
};

export const getAllCharacters = async (
  _: any,
  { limit, lastEvaluatedKey }: getAllArgs,
): Promise<PaginatedData<Character>> => {
  console.log(
    `[QUERY] getAllCharacters called with limit:${limit}, lastEvaluatedKey:${lastEvaluatedKey}`,
  );
  return CharacterDynamodb.Instance.getAllItems<Character>({ limit, lastEvaluatedKey });
};
