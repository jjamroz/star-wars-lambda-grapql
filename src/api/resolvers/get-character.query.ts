import { Character, GetDto } from '../../db/types';
import { CharacterDynamodb } from '../../db/character-dynamodb';

type getCharacterArgs = GetDto<Character>;

export const getCharacter = async (_: any, { id }: getCharacterArgs): Promise<Character | null> => {
  console.log(`[QUERY] getCharacter called with id ${id}`);
  return CharacterDynamodb.Instance.getItem<Character>({ id });
};
