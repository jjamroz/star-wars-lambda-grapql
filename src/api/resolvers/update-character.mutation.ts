import { CharacterDynamodb } from '../../db/character-dynamodb';
import { Character, UpdateDto } from '../../db/types';

type updateCharacterArgs = {
  input: UpdateDto<Character>;
};

export const updateCharacter = async (_: any, { input }: updateCharacterArgs) => {
  console.log(`[MUTATION] updateCharacter called with ${JSON.stringify(input, null, 4)}`);
  return CharacterDynamodb.Instance.updateItem(input);
};
