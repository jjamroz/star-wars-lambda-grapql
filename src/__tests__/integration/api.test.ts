import { typeDefs } from '../../api/schema';
import { resolvers } from '../../api/resolvers';
import { ApolloServer } from 'apollo-server-lambda';
import * as data from '../../../characters.json';
import { CHARACTERS } from '../queries';
import { Character } from '../../db/types';

jest.mock('aws-sdk', () => {
  return {
    DynamoDB: {
      DocumentClient: jest.fn(() => ({
        get: mockDynamoMethod({
          Item: data[0] as Character,
        }),
        put: mockDynamoMethod({
          Item: data[0] as Character,
        }),
        delete: mockDynamoMethod({
          Attributes: data[0] as Character,
        }),
        scan: mockDynamoMethod({
          Items: data as [Character],
          LastEvaluatedKey: {
            id: data[0].id,
          },
        }),
      })),
    },
  };
});

describe('API TESTS', () => {
  test('GET character', async () => {
    const server = new ApolloServer({
      typeDefs,
      resolvers,
    });

    const res = await server.executeOperation({
      query: CHARACTERS.GET_QUERY,
      variables: { id: 'SOME_VALID_ID' },
    });
    expect(res).toMatchSnapshot();
  });

  test('GET ALL characters', async () => {
    const server = new ApolloServer({
      typeDefs,
      resolvers,
    });

    const res = await server.executeOperation({
      query: CHARACTERS.GET_ALL_QUERY,
    });
    expect(res).toMatchSnapshot();
  });

  test('UPDATE character', async () => {
    const server = new ApolloServer({
      typeDefs,
      resolvers,
    });

    const res = await server.executeOperation({
      query: CHARACTERS.UPDATE_MUTATION,
      variables: {
        input: {
          id: data[3].id,
          patch: {
            name: 'GERENAL KENOOOOOBI',
            planet: 'HELLO THERE',
            episodes: ['NEWHOPE'],
          },
        },
      },
    });
    expect(res).toMatchSnapshot();
  });

  test('CREATE character', async () => {
    const server = new ApolloServer({
      typeDefs,
      resolvers,
    });

    const res = await server.executeOperation({
      query: CHARACTERS.CREATE_MUTATION_NO_ID,
      variables: {
        input: {
          name: 'GERENAL KENOOOOOBI',
          planet: 'HELLO THERE',
          episodes: ['NEWHOPE'],
        },
      },
    });
    expect(res).toMatchSnapshot();
  });

  test('DELETE character', async () => {
    const server = new ApolloServer({
      typeDefs,
      resolvers,
    });

    const res = await server.executeOperation({
      query: CHARACTERS.DELETE_MUTATION,
      variables: { id: 'SOME_VALID_ID' },
    });
    expect(res).toMatchSnapshot();
  });
});

const mockDynamoMethod = (returnValue) =>
  jest.fn().mockImplementation(() => {
    return {
      promise: jest.fn().mockResolvedValue(returnValue),
    };
  });
