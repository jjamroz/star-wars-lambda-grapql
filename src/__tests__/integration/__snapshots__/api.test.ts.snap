// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API TESTS CREATE character 1`] = `
{
  "data": {
    "createCharacter": {
      "episodes": [
        "NEWHOPE",
      ],
      "name": "GERENAL KENOOOOOBI",
      "planet": "HELLO THERE",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`API TESTS DELETE character 1`] = `
{
  "data": {
    "deleteCharacter": {
      "episodes": [
        "NEWHOPE",
        "EMPIRE",
        "JEDI",
      ],
      "id": "c74e4238-9ba8-4cc2-8c1c-eb9d20b76084",
      "name": "Luke Skywalker",
      "planet": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`API TESTS GET ALL characters 1`] = `
{
  "data": {
    "getAllCharacters": {
      "items": [
        {
          "id": "c74e4238-9ba8-4cc2-8c1c-eb9d20b76084",
          "name": "Luke Skywalker",
        },
        {
          "id": "a90fc28f-8b6a-4f08-a14b-b67c449fb8bd",
          "name": "Darth Vader",
        },
        {
          "id": "b4fc3e01-06e1-401d-91c0-f7410ca78005",
          "name": "Han Solo",
        },
        {
          "id": "67fb1ec0-57df-4461-9d5c-e077426f99fa",
          "name": "Leia Organa",
        },
        {
          "id": "7d78ad4b-cbe2-4d11-9e51-eeae26e0839a",
          "name": "Wilhuff Tarkin",
        },
        {
          "id": "df9976e4-20b1-4ed1-927e-18650f6fcef6",
          "name": "C-3PO",
        },
        {
          "id": "cb149869-47ec-4002-86fd-78c9106a556f",
          "name": "R2-D2",
        },
      ],
      "lastEvaluatedKey": "c74e4238-9ba8-4cc2-8c1c-eb9d20b76084",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`API TESTS GET character 1`] = `
{
  "data": {
    "getCharacter": {
      "episodes": [
        "NEWHOPE",
        "EMPIRE",
        "JEDI",
      ],
      "id": "c74e4238-9ba8-4cc2-8c1c-eb9d20b76084",
      "name": "Luke Skywalker",
      "planet": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`API TESTS UPDATE character 1`] = `
{
  "data": {
    "updateCharacter": {
      "episodes": [
        "NEWHOPE",
      ],
      "id": "c74e4238-9ba8-4cc2-8c1c-eb9d20b76084",
      "name": "GERENAL KENOOOOOBI",
      "planet": "HELLO THERE",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
